# Guided Project: Exploring Hacker News Posts

In this project I will ask the questions. Do Ask HN or Show HN recieve more comments on average? and Do posts created at certain time recieve more comments on average?

In [1]:
# read data
import csv
hn = open('hacker_news.csv')
hn = list(csv.reader(hn))
hn[:5]


[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'],
 ['12224879',
  'Interactive Dynamic Video',
  'http://www.interactivedynamicvideo.com/',
  '386',
  '52',
  'ne0phyte',
  '8/4/2016 11:52'],
 ['10975351',
  'How to Use Open Source and Shut the Fuck Up at the Same Time',
  'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/',
  '39',
  '10',
  'josep2',
  '1/26/2016 19:30'],
 ['11964716',
  "Florida DJs May Face Felony for April Fools' Water Joke",
  'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/',
  '2',
  '1',
  'vezycash',
  '6/23/2016 22:20'],
 ['11919867',
  'Technology ventures: From Idea to Enterprise',
  'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429',
  '3',
  '1',
  'hswarna',
  '6/17/2016 0:01']]

In [2]:
# removing headers from list 
headers = hn[0]
hn = hn[1:]
print(headers)
print(hn[:5])

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']
[['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01'], ['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '2', 'walterbell', '9/30/2015 4:12']]


In [3]:
# identifying the posts that begin with "ask hn", "show hn", and
# seprating the data in different lists
ask_posts = []
show_posts = []
other_posts = []

for post in hn:
    title = post[1]
    if title.lower().startswith("ask hn"):
        ask_posts.append(post)
    elif title.lower().startswith("show hn"):
        show_posts.append(post)
    else:
        other_posts.append(post)
print(len(ask_posts))
print(len(show_posts))
print(len(other_posts))
    

1744
1162
17194


In [4]:
# Calculate the average number of comments for "Ask HN 
total_ask_comments = 0

for post in ask_posts:
    total_ask_comments += int(post[4])
avg_ask_comments = total_ask_comments / len(ask_posts)
print(avg_ask_comments)



14.038417431192661


In [5]:
# Calculate the average nuber of comments for "Show HN posts

total_show_comments = 0

for post in  show_posts:
    total_show_comments += int(post[4])
avg_show_comments = total_show_comments / len(show_posts)
print(avg_show_comments)


10.31669535283993


On average total ask comments recieve 14 compared to 10 for total show comments. We will focus on the ask comments since there is more of a reponse on there comments.  

In [6]:
# calcuate the amount of Ask post and comments by hour
import datetime as dt

result_list = []

for post in ask_posts:
    result_list.append([post[6], int(post[4])])
    
counts_by_hour = {}
comments_by_hour = {}
date_format = "%m/%d/%Y %H:%M"

for each_row in result_list:
    date = each_row[0]
    comment = each_row[1]
    time = dt.datetime.strptime(date,date_format).strftime("%H")
    
    if time not in counts_by_hour:
        counts_by_hour[time] = 1
        comments_by_hour[time] = comment
    else:
        counts_by_hour[time] += 1
        comments_by_hour[time] += comment
print(comments_by_hour)   

{'17': 1146, '16': 1814, '02': 1381, '13': 1253, '11': 641, '14': 1416, '05': 464, '03': 421, '00': 447, '10': 793, '19': 1188, '04': 337, '15': 4477, '21': 1745, '01': 683, '08': 492, '22': 479, '06': 397, '12': 687, '18': 1439, '20': 1722, '09': 251, '07': 267, '23': 543}


In [7]:
# Calculate the average number of comments for Ask HN post by Hour

avg_by_hour = []

for hr in comments_by_hour:
    avg_by_hour.append([hr, comments_by_hour[hr] / counts_by_hour[hr]])
avg_by_hour

[['17', 11.46],
 ['16', 16.796296296296298],
 ['02', 23.810344827586206],
 ['13', 14.741176470588234],
 ['11', 11.051724137931034],
 ['14', 13.233644859813085],
 ['05', 10.08695652173913],
 ['03', 7.796296296296297],
 ['00', 8.127272727272727],
 ['10', 13.440677966101696],
 ['19', 10.8],
 ['04', 7.170212765957447],
 ['15', 38.5948275862069],
 ['21', 16.009174311926607],
 ['01', 11.383333333333333],
 ['08', 10.25],
 ['22', 6.746478873239437],
 ['06', 9.022727272727273],
 ['12', 9.41095890410959],
 ['18', 13.20183486238532],
 ['20', 21.525],
 ['09', 5.5777777777777775],
 ['07', 7.852941176470588],
 ['23', 7.985294117647059]]

In [8]:
# sorting and printing Values from list of lists

swap_avg_by_hour = []

for row in avg_by_hour:
    swap_avg_by_hour.append([row[1], row[0]])
    print(swap_avg_by_hour)

sorted_swap = sorted(swap_avg_by_hour, reverse=True)

print(sorted_swap)
    

[[11.46, '17']]
[[11.46, '17'], [16.796296296296298, '16']]
[[11.46, '17'], [16.796296296296298, '16'], [23.810344827586206, '02']]
[[11.46, '17'], [16.796296296296298, '16'], [23.810344827586206, '02'], [14.741176470588234, '13']]
[[11.46, '17'], [16.796296296296298, '16'], [23.810344827586206, '02'], [14.741176470588234, '13'], [11.051724137931034, '11']]
[[11.46, '17'], [16.796296296296298, '16'], [23.810344827586206, '02'], [14.741176470588234, '13'], [11.051724137931034, '11'], [13.233644859813085, '14']]
[[11.46, '17'], [16.796296296296298, '16'], [23.810344827586206, '02'], [14.741176470588234, '13'], [11.051724137931034, '11'], [13.233644859813085, '14'], [10.08695652173913, '05']]
[[11.46, '17'], [16.796296296296298, '16'], [23.810344827586206, '02'], [14.741176470588234, '13'], [11.051724137931034, '11'], [13.233644859813085, '14'], [10.08695652173913, '05'], [7.796296296296297, '03']]
[[11.46, '17'], [16.796296296296298, '16'], [23.810344827586206, '02'], [14.741176470588234

In [13]:
# sort the values and print the top 5 hours with the most ask comments
print("Top 5 Hours for 'Ask HN' Comments")

for avg, hr in sorted_swap[:5]:
    print("{}:{:.2f} average comments per post".format(dt.datetime.strptime(hr,"%H").strftime("%H:%M"), avg))

Top 5 Hours for 'Ask HN' Comments
15:00:38.59 average comments per post
02:00:23.81 average comments per post
20:00:21.52 average comments per post
16:00:16.80 average comments per post
21:00:16.01 average comments per post


The hour that recieve the most comments is at 15:00 time mark with 38.59 comments 60% percent more then from the highest to the lowest comment. 

# Conclusion

In this project, we analyzed ask posts and show posts to determine which type of post and time receive the most comments on average. Based on our analysis, to maximize the amount of comments a post receives, we’d recommend the post be categorized as ask post and created between 15:00 and 16:00 (3:00 pm est - 4:00 pm est).

However, it should be noted that the data set we analyzed excluded posts without any comments. Given that, it’s more accurate to say that of the posts that received comments, ask posts received more comments on average and ask posts created between 15:00 and 16:00 (3:00 pm est - 4:00 pm est) received the most comments on average.